primers/pcr/pcr.go

```diff
@@ -0,0 +1,138 @@
+				reverseLocations[reverseLocation] = append(reverseLocations[reverseLocation], primerIdx)
+			}
+		}
+	}
+
+	var forwardLocationInts []int
+	var reverseLocationInts []int
+	for key := range forwardLocations {
+		forwardLocationInts = append(forwardLocationInts, key)
+	}
+	for key := range reverseLocations {
+		reverseLocationInts = append(reverseLocationInts, key)
+	}
+	sort.Ints(forwardLocationInts)
+	sort.Ints(reverseLocationInts)
+
+	var pcrFragment string
+	var pcrFragments []string
+
+	// Next, iterate through the forwardLocations list
+	for i, forwardLocation := range forwardLocationInts {
+		// First, make sure that this isn't the last element in forwardLocations
+		if i+1 != len(forwardLocationInts) {
+			// If this isn't the last element in forwardLocations, then we can select the first reverseLocation that is less than the next forwardLocation
+			for _, reverseLocation := range reverseLocationInts {
+				if (forwardLocation < reverseLocation) && (reverseLocation < forwardLocationInts[i+1]) {
+					// If both are true, we have found the sequence we are aiming to PCR! Now, we get all primers from that forwardLocation and then
+					// build PCR fragments with each one.
+					for forwardPrimerIdx := range forwardLocations[forwardLocation] {
+						minimalPrimer := minimalPrimers[forwardPrimerIdx]
+						fullPrimerForward := primerList[forwardPrimerIdx]
+						for _, reversePrimerIdx := range reverseLocations[reverseLocation] {
+							fullPrimerReverse := transform.ReverseComplement(primerList[reversePrimerIdx])
+							pcrFragment = fullPrimerForward[:len(fullPrimerForward)-len(minimalPrimer)] + sequence[forwardLocation:reverseLocation] + fullPrimerReverse
+							pcrFragments = append(pcrFragments, pcrFragment)
+						}
+					}
+					break
+				}
+			}
+		} else {
+			foundFragment := false
+			for _, reverseLocation := range reverseLocationInts {
+				if forwardLocation < reverseLocation {
+					for forwardPrimerIdx := range forwardLocations[forwardLocation] {
+						minimalPrimer := minimalPrimers[forwardPrimerIdx]
+						fullPrimerForward := primerList[forwardPrimerIdx]
+						for _, reversePrimerIdx := range reverseLocations[reverseLocation] {
+							fullPrimerReverse := transform.ReverseComplement(primerList[reversePrimerIdx])
+							pcrFragment = fullPrimerForward[:len(fullPrimerForward)-len(minimalPrimer)] + sequence[forwardLocation:reverseLocation] + fullPrimerReverse
+							pcrFragments = append(pcrFragments, pcrFragment)
+						}
+					}
+					foundFragment = true
+				}
+			}
+			// If the sequence is circular and we haven't found a fragment yet, check the other side of the origin
+			if circular && !foundFragment {
+				for _, reverseLocation := range reverseLocationInts {
+					if forwardLocationInts[0] > reverseLocation {
+						// If either one of these are true, create a new pcrFragment and append to pcrFragments
+						for forwardPrimerIdx := range forwardLocations[forwardLocation] {
```

Comment:
This double nested for loop appears in 3 places. Consider refactoring into private helper function.